### PR TITLE
feat(stdlib): Vec type

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

This PR utilizes the recent additions of mutable references and slices to create a Vec type.

## Problem\*

This is work towards resolving #1556 

## Summary\*

I made a new type `struct Vec<T> { slice: [T] }` with a few builtin funcs that uses our slice builtins to implement a mutable vector type. 

If a user wants to update a slice they no longer have to use this syntax:
```
    let mut slice: [Field] = [];
    for i in 0..5 {
        slice = slice.push_back(i);
    }
    assert(new_slice.len() == 5);
```
and can instead do this:
```
    let mut vector: Vec<Field> = Vec::new();
    for i in 0..5 {
        vector.push(i);
    }
    assert(vector.len() == 5);
```
as the builtin Vec functions use mutable references. The Vec type is essentially a mutable wrapper over slices. 

## Documentation

- [X] This PR requires documentation updates when merged.

  

  - [X] I will submit a noir-lang/docs PR.

  

  - [X] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
> and can instead do this:
> 
```rs
let mut vector: Vec<Field> = Vec::new();
for i in 0..5 {
    vector.push(i);
}
assert(vector.len() == 5);
```
The type annotation is not necessary here, right?

File: crates/nargo_cli/tests/test_data_ssa_refactor/vectors/src/main.nr [hunk 1]
```diff
@@ -0,0 +1,29 @@
+use dep::std::collections::vec::Vec;
+
+fn main(x : Field, y : pub Field) {
+
```
Comment (jfecher) [inline crates/nargo_cli/tests/test_data_ssa_refactor/vectors/src/main.nr:4]:
```suggestion
fn main(x: Field, y: pub Field) {
```

File: crates/nargo_cli/tests/test_data_ssa_refactor/slices/src/main.nr [hunk 2]
```diff
@@ -5,7 +5,7 @@ fn main(x : Field, y : pub Field) {
     /// is not yet supported
 
     let mut slice: [Field] = [0; 2];
-
+
```
Comment (jfecher) [inline crates/nargo_cli/tests/test_data_ssa_refactor/slices/src/main.nr:8]:
```suggestion

```

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs [hunk 3]
```diff
@@ -375,12 +375,16 @@ impl Context {
...
             if let Some(var) = self.current_side_effects_enabled_var {
+                // Ignore the error if side effects are disabled.
                 if self.acir_context.is_constant_one(&var) {
                     // TODO: Can we save a source Location for this error?
                     panic!("Index {} is out of bounds for array of length {}", index, array.len());
                 }
+            } else {
+                // If `current_side_effects_enabled_var` are `None` it means
+                // that we are in the main function and want to panic
+                panic!("Index {} is out of bounds for array of length {}", index, array.len());
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:387]: Can we revert this change in favor of #1906? I'd rather remove the `Option` entirely.

File: noir_stdlib/src/collections/vec.nr [hunk 4]
```diff
@@ -0,0 +1,63 @@
+struct Vec<T> { 
+    slice: [T]
+}
+
+// These methods are all stubs currently and aren't implemented internally yet.
+// For a similar reason, no constructor for Vec is exposed yet since the type
+// is still in-progress.
```
Comment (jfecher) [inline noir_stdlib/src/collections/vec.nr:7]:
```suggestion
// A mutable vector type implemented as a wrapper around immutable slices.
// A separate type is technically not needed but helps differentiate which operations are mutable.
```

File: noir_stdlib/src/collections/vec.nr [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+impl<T> Vec<T> {
+    fn new() -> Self {
+        Self { slice: [] }
+    }
+
```
Comment (jfecher) [inline noir_stdlib/src/collections/vec.nr:12]:
```suggestion

    // Create a Vec containing each element from the given slice.
    // Mutations to the resulting Vec will not affect the original slice.
    fn from_slice(slice: [T]) -> Self {
        Self { slice }
    }

```

**Comment by vezenovm** [comment]
> The type annotation is not necessary here, right?

Yeah it is unneeded I removed

**Review by jfecher — COMMENTED** [review]
Looks good. Once we fix slices in if statements we can add functions like `map` as well.

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs (hunk 3 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:387]: yup will revert

Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs]: done

**Review by jfecher — APPROVED** [review]
👍
